Hpc2 theory (#27)
ARCTraining/hpc2-software: book/course/theory.md

# Theory

## Introduction

### Why

This lesson provides supporting information to the tools used to use, build and maintain software on the Universities HPC systems by introducing you to the terminology and technologies that will be covered by further lessons in this course.

### Overview of topics

* Linux ????
* Build Tools
* Package Managers
* Platforms

### Types of files (WHERE DOES THIS FIT ???????????????)

There are numerous file types you need to be aware of when using and developing code for HPC.
Below are the key types of files:

* executables - these types of files contain compiled binary code. When opened, the operating system will execute the instructions of the code to run the program.

> * Libraries - these files usually contain compiled code, but cannot be directly run (they can also contain data, templates, documentation). They provide interfaces to be accessed from an existing running program. Libraries are useful for sharing common functionality between different programs.

Reviewer: Not sure I like this definition, although I think the additional note that the term often gets used to cover other things is valid.  I think in terms of building software, I'd go with more of a:

A Library in Linux
A library is a collection of pre-compiled pieces of code called functions. The library contains common functions and together, they form a package called — a library.

Or maybe:

In programming, a library is an assortment of pre-compiled pieces of code that can be reused in a program.

### Software locations

* (standard linux) locations
* library paths
* user locations

### Installing software

Software usually requires installation before it can be run.

* executable installer
* package file install
* archive with instructions of how to extract and install and run.
* Source archives with build instructions and scripts.

```{Note}
It's possible to have an executable program or script that can be run directly without installation.
```

Executable installers, may not include dependancies (other files required to run), these could be provided by default by the Operating System or may require installing seperatly.

### Package Managers

Package managers are used to manage applications, tools and libraries as packages.

Packages are archives (often compressed files like zip) and can contain any combination of software, libraries and their metadata.

Packages are collected and maintained together as distributions (eg conda-forge).

Packages managers keep track of the packages (and their versions/varients) available, installed and their dependencies.

  they usually allow the following functionality:

> 

Author: I agree with @jhodrien on this, in bullet points full sentences should be punctuated.

* Make - the traditional Unix/Linux build tool.

### Compilers

A compiler is a program which translates source code written in a programming language into target code. This is typically into machine code to create an executable program.

ARC3 and ARC4 have the [Intel](https://arcdocs.leeds.ac.uk/software/compilers/intel.html) compiler set as default, with the [GNU](https://arcdocs.leeds.ac.uk/software/compilers/gnu.html) and [LLVM](https://arcdocs.leeds.ac.uk/software/compilers/llvm.html) compilers being popular alternatives available on the module system.

#### Interpreters

An interpreter reads and executes source code directly. They generally runs code slower than a compiled executable as it has to compile code every time the code is run, but there are optimisations such as bytecode compilation and caching that can vastly improve performance.

Python is a popular interpreted language and the most common interpreter is the default, [CPython](https://en.wikipedia.org/wiki/CPython).

Some languages are capable of being compiled or run through an interpreter.

### Build Configuration

> Tools like CMake can be used to create different platform specific for different build tools configurations.

Reviewer 2: ```suggestion
Tools like CMake can be used to create different platform specific for different build tools configurations.
```
This doesn't make sense, think you're missing a word after "different platform specific"

Configuration files usually contain locations of the source code, output locations, of existing components (libraries, assets such as help documentation, data).

They can contain options for conditional builds such as for debug and release builds and for additional features to take advantage of optional components such as input/output format/interfaces or hardware - such as GPUs.

### Packaging

```{Note}

> Users do not have sufficent permissions to install system packages on our HPC systems.

Reviewer 2: ```suggestion
Users do not have sufficient permissions to install system packages on our HPC systems.
```

Containers provide a mechanism for users to install system packages in a user managed environment.
```

> Linux and package distribtions usually provide tools to automate many steps of the process of creating software packages from source; this will include steps to compile source code, package the files required and test the package is installable/removable.

Reviewer 2: ```suggestion
Linux and package distributions usually provide tools to automate many steps of the process of creating software packages from source; this will include steps to compile source code, package the files required and test the package is installable/removable.
```

### Distribution

There is usually little need to distribute built academic code.

> If you want to distribute your source code, you can use online repositories such as GitHub. These can be linked into open research repositories like Zenodo.

Reviewer 2: ```suggestion
If you want to distribute your source code, you can use online repositories such as [GitHub](https://github.com/). These can be linked into open research repositories like [Zenodo](https://zenodo.org/).
```

Containers (and their definitions) are a popular way of distributing ready to run environments.

## Platforms

The platforms mentioned below provide different levels of abstraction to run your software and code in.  These include:

* Machines (physical and virtual)
* Containers
* Virtual Environments

They are extremely useful for helping make your projects reproducible, and also maintain compatibility and consistency for your projects.

### Physical Machines

This refers to your physical devices such as laptops, PCs, phones, tablets.

They utilise an **Operating System** (eg Windows, OSX, Android Linux) to control the hardware resources (CPU, storage, memory) to access data and run software and your code.

### Virtual Machines

```{Note}
ARC3/ARC4 cannot run virtual machines.
```

Virtual machines refer to applications (and system level hypervisors) that provide a software representation of hardware to run guest machines.

It is possible to create and run multiple guest machines on a single host.

The guest machines can use different virtual hardware to the host phyiscal machine and any operating system that supports the virtual hardware.

### Containers

> Containers are an Operating System level virtualisation that provides isolated instances.

Reviewer 2: ```suggestion
Containers are an operating system level virtualisation that provides isolated instances.
```

They utilise the host operating sytem's kernel and access to (usually a limited set of) hardware resources (memory, CPU, storage, input devices).

> Containers have less overheard (often no noticable performance loss) due to not needing to emulate any hardware, more flexible and require less storage and configuration.

Reviewer: ```suggestion
Containers have less overhead (often no noticeable performance loss) due to not needing to emulate any hardware, are more flexible and require less storage and configuration.
```

ARC3 and ARC4 support Singularity and Apptainer (the new successor to Singularity) container tools. They support the import and running of Docker containers.

### Virtual Environments

Virtual Environments are environments for managing isolated sets of user software (tools and libraries).
These are usually managed by a tool such as Anaconda or virtualenv.

> Each environemt created can have its own set of software and when combined with package management can specify specific (or minimum or maximum) versions of some (or all) applications and libraries.

Reviewer 2: ```suggestion
Each environment created can have its own set of software and when combined with package management can specify specific (or minimum or maximum) versions of some (or all) applications and libraries.
```

> ARC3 and ARC4 provides Anaconda as the primary tool for managing environments, although others (such as virtualenv for Python) are available or installable.

Reviewer 2: ```suggestion
ARC3 and ARC4 provides the Anaconda module (that contains `conda` as the primary tool for managing environments, although others (such as virtualenv for Python) are available or installable.
```

> ARC3 and ARC4 provides Anaconda as the primary tool for managing environments, others (such as virtualenv for python) as available or installable.

Reviewer: ```suggestion
ARC3 and ARC4 provides Anaconda as the primary tool for managing environments, although others (such as virtualenv for Python) are available or installable.
```
